### DISCUSSION TITLE
ESP32: change partitioning via OTA?

### DISCUSSION THREAD

**Discussion by msperl** [arendst/Tasmota#12246] [Q&A] [answered]
Hi!

Is there a way to change the partitioning of a tasmota device via some OTA mechansims? I.e: a special tasmota image that needs to get installed.

My use case Is: I have a lanbon l8 device which I have flashed with a standard partitioning scheme of 4MB, but which has actually 8MB Flash.
So I am now wondering how I can change the partitioning without having to go back to the serial console flashing.

Note that it is OK if I loose the content of the filesystem itself.

**Comment by barbudor** [comment]
OTA only update the firmware block 
I'm afraid that to change the partition scheme you need to serial flash the device

**Comment by Jason2866** [comment]
Not possible with standard OTA procedure as @barbudor already wrote. A changed (correct) partitions.bin needs to be written at flash address 0x8000. This would need extra code in Tasmota, beside the aspect it needs special care, a false/defect partions.bin will result in a non starting device. Only a serial flash can recover.
More infos https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/partition-tables.html

**Comment by sfromis** [comment]
Only chance would be if someone (like you) have created special firmware for doing repartitioning without erase, and in that case flash Tasmota again after running that. Not seen such, not searched either.

**Comment by Jason2866** [comment]
Mhh, just changing the SPIFFS (used partition for Tasmota Ufilesystem) partition size would be possible since it is the last partition and no changes are needed to the other partitions. Still extra code needed to do this.

**Reply by msperl** [reply]
it would be just like raspberry os images that increase the partition and filesystem on first use.
Not sure how much code is needed to implement that.

And yes: I understand that I need to go back to the serial flashing at this very moment.

**Reply by msperl** [reply]
but looking into it I now also find that there are no partitions.bin that use the other 4MB for SPIF.
Need to look into how to create one now...

**Reply by Jason2866** [reply]
See here our partition.csv for the esp32 with 16M flash https://github.com/arendst/Tasmota/blob/development/esp32_partition_app2944k_spiffs10M.csv

**Reply by Jason2866** [reply]
It is easy, just calculate the sizes you want. Usage is defined in a `[env]` https://github.com/arendst/Tasmota/blob/development/platformio_tasmota_env32.ini#L50-L59
Do a Lanbon env, just add in file `platformio_tasmota_cenv.ini` following 
```
[env:tasmota32-Lanbon]
extends                 = env:tasmota32_base
build_flags             = ${common32.build_flags} -DFIRMWARE_TASMOTA32
board_build.partitions  = esp32_partition_appxxxxk_spiffsxxM.csv
```

**Reply by barbudor** [reply]
I did this one for the ESP32 D4 which has 8MB embedded flash
[esp32_partition_app1984k_spiffs4M.csv](https://github.com/arendst/Tasmota/files/6571877/esp32_partition_app1984k_spiffs4M.csv.txt)
(rename the file as .csv)

**Comment by s-hadinger** [comment]
Changing partition within Tasmota would require to be VERY careful, chances of bricking are high.

**Reply by msperl** [reply]
So what is the risk?

The update fails and bricks the device because I decide to change the partition and want to avoid doing that via the serial cable.
Rescue: update/erase/flash via the serial cable to unbrick.

In both cases I have to reattach the serial cable (if it does not work), but if it works I avoid doing that with 30 device and there is no longer the need to maintain those special scripts for m5 or odroid (as long as people are happy with the default split for the firmware itself, but I am talking here about people who want to use the precompiled images not something selfcompiled where you then can make all the choices of your own)

in both cases I would also loose all my settings.

Note that I am just asking  for an automatic spif increase (=last partition) on first install not a full partition table reorganization…

but, yes: in principle a flasher that can reliably change the partition table completely ota would be nice -  but it is not really necessary!

**Reply by s-hadinger** [reply]
The risk is more on users that don't fully understand what this feature is doing and would try it blindly. If you really want to change the partition, I agree that's more an opportunity than a risk.

Let me see if can just do a Berry extension to flashing. This would prevent from adding any specific code for this.

**Reply by barbudor** [reply]
If the idea is only to grow the file system to fill-up the flash without changing the other blocks, may be that could be done in @Jason2866's ESP-Flasher too ?
When flashing, identify the size of flash and adapt the partition so LittleFS block fills up the rest of the flash ?
That would still not cover changing dynamically after initial flashing but that could make if easy to use for many users ?

**Reply by msperl** [reply]
well I think that would be a huge improvement!
If it is Berry or something inside the flasher is less important (but flasher does not work on the commandline).
But I guess making this increase of the partition transparent during flashing/startup would make peoples lives easier.

**Reply by Jason2866** [reply]
> If the idea is only to grow the file system to fill-up the flash without changing the other blocks, may be that could be done in @Jason2866's ESP-Flasher too ?
> When flashing, identify the size of flash and adapt the partition so LittleFS block fills up the rest of the flash ?
> That would still not cover changing dynamically after initial flashing but that could make if easy to use for many users ?

The source code of ESP-Flasher is [here](https://github.com/Jason2866/ESP_Flasher) and the tool to generate the partition.bin is [here](https://github.com/espressif/arduino-esp32/blob/master/tools/gen_esp32part.py)

**Reply by Jason2866** [reply]
If someone is adopting ESP-Flasher by recognizing bigger flash size, it should increase the OTA partitions to the size which is used for Core2 and Odroid-go too (2944k) and adopt the SPIFFS size partition accordingly

**Reply by s-hadinger** [reply]
I have checked what is doable with Berry. Unfortunately esp-idf prevents from writing the partition sector from the application, unless compiled with `CONFIG_SPI_FLASH_DANGEROUS_WRITE_ALLOWED`

I don't see any quick solution, except having a customer esp-idf with this option enabled. @Jason2866 would it be possible?

**Reply by Jason2866** [reply]
Possible, will do if github is back to normal status. At the moment github actions are non working.
Edit: github actions are working again.
Framework with the changed option `CONFIG_SPI_FLASH_DANGEROUS_WRITE_ALLOWED` can be used with Platformio [env] https://github.com/arendst/Tasmota/blob/development/platformio_tasmota_cenv_sample.ini#L28-L37

**Comment by Jason2866** [comment, marked as answer]
Will refactor/rework platformio setup for ESP32 devices with different flash sizes (4M, 8M, 16M).
So it will be possible to compile Tasmota32 for different flash sizes just by selecting a other tasmota32 env.
So it is easy for everyone to self compile and flash this variants via VSC. **VSC/Platformio takes care of everything and flashes all needed files correctly**
Edit: PR #12257

**Reply by s-hadinger** [reply]
Thanks Jason.
I have pushed #12279 that introduces `import flash` module in Berry. So we can now start playing around.
Parsing and changing partition data seems more feasible now.

**Comment by s-hadinger** [comment]
#12310 Introduced Berry code to read/change/write back partition table.

@msperl You should have everything you need to change on the fly the partition table.

**Reply by msperl** [reply]
Thanks, I guess I still need to understand the fine details of how partitioning works in detail to come up with something…

**Reply by msperl** [reply]
Maybe 3 questions:
* is there a means in Berry to identify from which app partition we are currently running?
* is it safe to say that berry does not mmap the  (bytecode) scripts/libraries into memory (or does an execute in place), which would wreck harvoc when overwriting/erasing the partition
* is there a way to get the "Program Size" value displayed on the http information page in Berry (mostly to check that a resize of an app partition would not truncate the app itself - typical use case would be increasing the size, so it is not an absolute need)?

Background thoughts:
* if we are booting from APP1 we should be easily be able to increase the size of APP1 and move/increase APP2 and move/decrease SPIFF accordingly (erasing it just in case - maybe just a few blocks are needed to get deleted)
* if we are booting from APP1 and the Program size is smaller than the target size, then it should also be possible to reduce the partition APP1 and APP2 while increasing SPIFF
* If we are booting from APP2 we need first to move to APP1 - i.e: we need to flash tasmota and reboot again for a switch of the active partition - but maybe APP2 is just used for the temporary OTA installer, so it is never an issue when using tasmota (but I have no idea about the fine details of ESP32, booting and OTA)

**Reply by sfromis** [reply]
Not very elegant, but doable:
```
tasmota.add_rule("StatusMem#ProgramSize", / value,trigger,map -> print(value))
tasmota.cmd("Status 4")
tasmota.remove_rule("StatusMem#ProgramSize")
```

**Reply by s-hadinger** [reply]
Good questions. I updated `partition.be` to be complete (I hope), so you can look at details, change them and write back. It also works to change the active OTA partition.


> * is there a means in Berry to identify from which app partition we are currently running?

Now yes, use `p.otadata`
Example: `<instance: Partition_otadata(ota_active:0, ota_seq=[5,4], ota_max=1)>`
The important part is `ota_active:0` meaning that the active partition is `ota0`. The rest are details.

> * is it safe to say that berry does not mmap the  (bytecode) scripts/libraries into memory (or does an execute in place), which would wreck harvoc when overwriting/erasing the partition

Yes, Berry runs from Flash. Erasing and writing to the current partition would require code in IRAM.

> * is there a way to get the "Program Size" value displayed on the http information page in Berry (mostly to check that a resize of an app partition would not truncate the app itself - typical use case would be increasing the size, so it is not an absolute need)?

That's still missing...

> Background thoughts:
> 
> * if we are booting from APP1 we should be easily be able to increase the size of APP1 and move/increase APP2 and move/decrease SPIFF accordingly (erasing it just in case - maybe just a few blocks are needed to get deleted)

I think you mean `app0` the first app partition. However I don't know the behavior of SPIFFS when you change the partition size...
But I see where you are heading, and yes, it could allow for some creativity.

> * if we are booting from APP1 and the Program size is smaller than the target size, then it should also be possible to reduce the partition APP1 and APP2 while increasing SPIFF
> * If we are booting from APP2 we need first to move to APP1 - i.e: we need to flash tasmota and reboot again for a switch of the active partition - but maybe APP2 is just used for the temporary OTA installer, so it is never an issue when using tasmota (but I have no idea about the fine details of ESP32, booting and OTA)

**Reply by s-hadinger** [reply]
There is now `tasmota.memory()

Example: `{'frag': 51, 'program_free': 1856, 'flash': 4096, 'heap_free': 226, 'program': 1679}`

Getting Flash size: `tasmota.memory['flash']`

**Comment by s-hadinger** [comment]
Reviving this topic, please check #12412, this is a first version of Partition Manager. For now it is read-only.

**Reply by msperl** [reply]
i have just tried it now (got a bit of time at hand), but:
a) I do not find all the 2 berry files as of now (the manager file ist not in the latest development branch any longer - especially the partition_manager.be is missing)
b) if I gather it from the original diff (commit 47db4f360c30881593f9aeb18ff16fcf37123c6) I get an error when I run the import: ```BRY: Exception> 'syntax_error' - input:1: expected 'as' before 'EOS'``` (using todays build: 9.5.0.2(tasmota) 2021-07-11T07:43:26)
So is there something missing in the latest build? - Maybe it is on purpose?

**Reply by s-hadinger** [reply]
It's much simpler now. Just copy `partition.be` or `partition.bec` in the file system and type `import partition` in the Berry console

**Reply by s-hadinger** [reply]
Go back to the Console menu and you have a new button

**Reply by msperl** [reply]
ah - that explains a lot (also why there is a bec version included in git - I had to use that one already to increase the SPIF again...)

Thanks for the effort.

**Reply by s-hadinger** [reply]
'bec' is just precompiled Berry code, which is sometimes smaller
